### ISSUE TITLE
Undefined terms need an optional default expansion (e.g., via `@vocab` in second `@context`)

### ISSUE THREAD

**Issue by OR13** [w3c-ccg/traceability-vocab#344]
```
_:c14n0 <https://w3id.org/traceability/#undefinedTermbar> "baz 123 123123123" .
```

should look like:

```
_:c14n0 <https://w3id.org/traceability/#undefined-term-bar> "baz 123 123123123" .
```

or something more readable.

**Comment by nissimsan** [comment]
Agreed, `https://w3id.org/traceability/#undefinedTermbar` is quite annoying to look at.

**Comment by TallTed** [comment]
@OR13 -- Change issue title from `exampansion` to `expansion`?

**Comment by OR13** [comment]
Consider also:

```
https://w3id.org/traceability/undefined-term/#bar ?
```

**Comment by TallTed** [comment]
A "temporary" or "stop-gap" `https://w3id.org/traceability/undefined-term/#bar` will stick ... and if is defined later ... it will be a _defined undefined-term_.

**Comment by OR13** [comment]
^ @TallTed yes, thats exactly my point!... not sure if thats a thing other folks have done.

**Comment by TallTed** [comment]
@OR13 -- My points are that —

1. Attaching a definition to a term which includes a string like `undefined-term` in its URI literal will (eventually) be problematic, even though the plain-language meanings of strings in URI literals are generally supposed to be ignored by humans.
2. Introducing a term like `https://w3id.org/traceability/undefined-term/#bar` will not make it easy to change to a term like `https://w3id.org/traceability/#bar` once a definition is settled on.

In other words, I don't think the `https://w3id.org/traceability/undefined-term/#bar` idea is a good one to act on.

**Comment by OR13** [comment]
@TallTed what is your recommended change to `@vocab` to address undefined terms that are currently assigned any JSON that was built by someone who does not understand JSON-LD?

Are you saying your are in favor of removing `@vocab` and throwing errors instead?

**Comment by OR13** [comment]
For context:

```
"@context": {
    "@vocab": "http://schema.org/",
    "goog": "http://schema.googleapis.com/",
    "resultScore": "goog:resultScore",
    "detailedDescription": "goog:detailedDescription",
    "EntitySearchResult": "goog:EntitySearchResult",
    "kg": "http://g.co/kg"
  },
```

- https://developers.google.com/knowledge-graph

**Comment by TallTed** [comment]
[@OR13]
> @TallTed what is your recommended change to `@vocab` to address undefined terms that are currently assigned to any JSON members that were built by someone who does not understand JSON-LD?
>
> Are you saying your are in favor of removing `@vocab` and throwing errors instead?

Possibly. I have not heard nor read solid arguments on either side of the "removing `@vocab`" option; I expect both exist.

This issue does not appear to be focused on the possibility of "removing `@vocab`", nor on any "recommended change to `@vocab` to address undefined terms that are currently assigned to any JSON members that were built by someone who does not understand JSON-LD".

This issue *appears* only to be focused on `undefinedTermbar` being somewhat less readable than `undefined-term-bar` (or the unsuggested `undefinedTermBar`), with which I do not disagree, but that does not logically lead me, nor I think anyone else, to discussion as in your quoted comment above.

There is no discussion here of how `undefinedTermbar` came about in the first place, nor why `undefined-term-bar` makes more sense than the unsuggested `undefinedTerm-bar`, nor what would make `undefined-term/#bar` a good idea (which I continue to think it is not, but I would consider reasoned argument before condemning it entirely).  Adding any (or preferably all) of these ingredients would make this issue more coherent and, I hope, addressable.

**Comment by OR13** [comment]
Happy to give some background.

When you don't include an `@vocab` in your context, all terms with no definition cause JSON-LD errors.

Many developers who are not familiar with JSON-LD find this frustrating.

Including `@vocab` "hides" these errors behind expanded IRIs.

We chose to signal that any IRI expanded by `@vocab` was essentially a "warning" ... so it doesn't error, but it also doesn't provide nice semantic data.... and this really hurts your when you start integrating with other graph technologies, such as SPARQL or CYPHER.

I think it's best that we keep `@vocab` and that we strive to eliminate its use.

Ideally when you follow a term definition that was expanded using `@vocab` you should see a gentle warning like:

"This term was not defined properly in a context, and therefore interoperability and advanced semantic feature of verifiable credentials might not be available, please consider contacting the issuer and advising them to define this term better using an existing industry ontology".

**Comment by TallTed** [comment]
I agree with [all this](https://github.com/w3c-ccg/traceability-vocab/issues/344#issuecomment-1205813908).

I don't see how any of it flows from the issue title, "Fix undefined term expansion", nor from the [initial comment](https://github.com/w3c-ccg/traceability-vocab/issues/344#issue-1167343790), nor from nor to much of anything else being said here.

Maybe this issue should just be closed, until a clearer formulation can be concocted for a new issue's title and initial comment.

**Comment by brownoxford** [comment]
Someone please correct me if I am wrong here, but as I understand it, `@vocab` is used specifically to declare a default vocabulary from which all terms derive, without having to declare specific mappings for each term.

> A "temporary" or "stop-gap" https://w3id.org/traceability/undefined-term/#bar will stick ... and if is defined later ... it will be a defined undefined-term.

I believe that if a previously undefined term is later defined, it will not longer be part of the default vocabulary specified by `@vocab`, and will therefore no longer have `undefined-term` in the URI as @TallTed was mentioning here.

**Comment by OR13** [comment]
Thats right. We have 2 objectives:

1. use `@vocab` to make JSON-LD less painful to noobs
2. allow expert to fix errors made by noobs.

**Comment by TallTed** [comment]
Oh, so you don't mean to define `https://w3id.org/traceability/undefined-term/#bar`; you mean to use the surfacing of that URI to prompt definition of `https://w3id.org/traceability/bar`?

That makes much more sense, and I have no objection to it.

**Comment by TallTed** [comment]
I stand by my assertion that this issue should have both its title and initial comment (explanation/detail of title) edited to more accurately reflect the problem that is being raised, and the solution being proposed thereto.

**Comment by OR13** [comment]
@TallTed propose some language :)

As of right now, I am in favor of doing this:

```
"@vocab": "https://w3id.org/traceability/undefined-term/#"
```

This would yield:

```
https://w3id.org/traceability/undefined-term/#bar
```

For cases where a JSON-LD noob forgets to define `bar` properly.

and would allow for both:

```
https://w3id.org/traceability/bar
https://w3id.org/traceability#bar
```

In the case they know what they are doing.

**Comment by nissimsan** [comment]
If VC2.0 did this, we wouldn't need to...

**Comment by TallTed** [comment]
Suggested title revision:
```Undefined terms need an optional default expansion (e.g., via `@vocab` in second `@context`), related to https://github.com/w3c/vc-data-model/issues/953```

**Comment by mfosterio** [comment]
What are your thoughts of describing the undefinedTerm in Schema.org with the type PropertyValue?
{
  "@context": "http://schema.org/",
  "@type": "PropertyValue",
  "name" : "bar",
  "value": "baz"
}

**Comment by mfosterio** [comment]
What are your thought of referencing the schema.org type PorpertyValue to generate something like the following"
{
  "`@type`": "https://schema.org/Organization",
  "https://schema.org/description": "Realigned maximized alliance",
  "https://schema.org/name": "Bartell Inc 🔥",
  "https://w3id.org/traceability/#undefinedTerm": {
    "`@type`": "http://schema.org/PropertyValue",
    "http://schema.org/value": "baz",
    "https://schema.org/name": "bar"
  }
}

**Comment by OR13** [comment]
See https://github.com/w3c/vc-data-model/blob/main/vocab/credentials/v2/issuer-dependent.html

**Comment by OR13** [comment]
https://github.com/w3c/vc-data-model/blob/main/contexts/credentials/v2#L4

**Comment by OR13** [comment]
```
    "@vocab": "https://www.w3.org/ns/credentials/issuer-dependent#",
```

**Comment by OR13** [comment]
We should do what the VCWG is doing, we should start by using their vocab.

**Comment by nissimsan** [comment]
No disagreement on the call to go with @OR13's suggestion above. 
https://github.com/w3c-ccg/traceability-vocab/issues/344#issuecomment-1503837575

**Comment by OR13** [comment]
Make sure to also update the reporting feature that renders the charts.
